### ISSUE TITLE
Split out d3fc-extent

### ISSUE THREAD

**Issue by chrisprice** [ScottLogic/d3fc#994]
Split out extent and eliminate a few of its "features" -
* Remove support for field names `fields(['x'])` - equivalent to `fields([d=>d.x])`
* Remove support for 1 level of nested arrays - it should either support recursion through n levels or no recursion.
* Split out date functionality e.g. `import {linearExtent, dateExtent} from 'd3fc-extent'` 
* Force pad being a tuple.
* Remove `include([n])` - equivalent to `fields([d=>n])`

**Updated 22-06**

**Comment by rhewitt** [comment]
Hi, Is this the field-name support that's being eliminated ```fc.util.extent().fields(["f1","f2"])```? Asking because it's one of my favorite features. Will be sad to see it go :-(

**Comment by chrisprice** [comment]
Thanks for the feedback @rhewitt. Just so that it's clear what I'm proposing, here's an example usage -
```js
import { linearExtent } from 'd3fc-extent';
const xExtent = linearExtent()
  .fields([d=>d.f1, d=>d.f2]);
```

*When typing that realised that `linear` would be an awkward name for the export as you'd always end up aliasing it. So I've embraced the naming convention from [d3-scale](https://github.com/d3/d3-scale).*

Using the new arrow function syntax, each field definition ends up being 3 characters longer to type out with the associated noise of the `d` (de-)reference.

**Comment by rhewitt** [comment]
Hi Chris, Hey thanks for replying. I'm not really a javascript programmer; I just hack with it, so maybe I'm missing or misunderstanding something. Isn't the d=> syntax just a functional? It's dereferencing in this example, but couldn't this function be doing something more dynamic? If that's correct, this seems like a useful functionality that it would be real a shame to lose. I could imagine using it, for example, to specify a virtual field to implement, say, a 5% margin. I could also imagine embedding a conditional. What do you think...or have I misunderstood the JS? 

Looks like this is the ```mutatedFields``` in extent.js. Is that correct?

Note: It may not matter, but I didn't understand the part about aliasing linear and why that would be necessary.

**Comment by chrisprice** [comment]
> Isn't the d=> syntax just a functional?

Yep, just ES6 syntactic sugar for a `function` (with a lexically bound context).

> It's dereferencing in this example, but couldn't this function be doing something more dynamic? If that's correct, this seems like a useful functionality that it would be real a shame to lose. I could imagine using it, for example, to specify a virtual field to implement, say, a 5% margin ([d=>0.95*d.x,d=>1.05*d.x]). I could also imagine embedding a conditional. What do you think...or have I misunderstood the JS?

You've hit on my preference here. I think the `d=>d.x` arrow function syntax allows the existing functionality of specifying fields, adds the ability to specify virtual fields and all for expense of a few extra characters. I personally also like how it removes the *magic* of just specifying field names.

> Looks like this is the mutatedFields in extent.js. Is that correct?

Yes `mutatedFields` is a projection of `fields` applied when it detects that a given `field` is a `string` rather than a `function`.

> Note: It may not matter, but I didn't understand the part about aliasing linear and why that would be necessary.

I was just referring to the naming of the new component. I now have a preference for `linearExtent` so that you can just write -

```js
import { linearExtent } from 'd3fc-extent';
const xExtent = linearExtent()
  .fields([d=>d.f1, d=>d.f2]);
```

Instead of -

```js
import { linear as linearExtent } from 'd3fc-extent';
const xExtent = linearExtent()
  .fields([d=>d.f1, d=>d.f2]);
```

**Comment by ColinEberhardt** [comment]
Hi @rhewitt - we're all just hackers ;-)

Regarding the `d=>` syntax, it is an arrow function. In this context it can simply be used as a more shorthand mechanism for creating a function.

The following are all equivalent:

functions:

```
const yExtent = linearExtent()
    .fields([function(d) {
       return d.high;
    }, function(d) {
       return d.low;
    }]);
```

arrow functions:

```
const yExtent = linearExtent()
    .fields([
      (d) => { return d.high; },
      (d) => { return f.low; }
    ]);
```

arrow functions, in their most compact form:

```
const yExtent = linearExtent()
    .fields([d => d.high, d => d.low]);
```

As you can see, with arrow functions the syntax is really quite concise. And yes, as you observed, you can do things like `d => d.high * 0.6`.

The current API which allows you to specify property names is ever so slightly more concise:

```
const yExtent = linearExtent()
    .fields(['high', 'low']);
```

However, it is pretty weak. You can't do maths, `d => d.high * 0.6`, navigate properties, `d => d.boll.upper`, you don't get any static analysis from your listing tools, or type checking with TypeScript.

Basically you lose more than you gain.

The only issue is that arrow functions are not fully supported in all browser, so this does mean you will have to transpire your code (e.g. babel) to benefit from this. But with d3fc we are looking to the future rather than the past.

**Comment by ColinEberhardt** [comment]
@chrisprice - looks like we need a ticketing system!

**Comment by chrisprice** [comment]
To save anyone else typing the words into google - [caniuse arrow functions](http://caniuse.com/#feat=arrow-functions).

**Comment by rhewitt** [comment]
>The current API which allows you to specify property names is ever so slightly more concise....You can't do maths, ```d => d.high * 0.6```, navigate properties, ```d => d.boll.upper```....

Hi @ColinEberhardt and thanks for adding your thoughts. I'm a little baffled by the comment quoted above, though, because you certainly can do these things with the current API. The ```mutatedFields``` variable is an array mapping of a function. If you pass a functional to extents, it's mapped to that functional. If you pass a string, it maps to an accessor function that returns the named field. To validate that the current API works with functionals that include math, I took the [gist I made](https://gist.github.com/rhewitt/3a170d58d8eff40fc1b858e1d9b1124d) for @Micjoyce and changed it to use functionals in the X and Y extents instead of field names : 
```
var chart = fc.chart.cartesian(
   d3.scale.linear(),
   d3.scale.linear())
   .margin({bottom: 30, right: 40, left: 8})
   .xDomain(fc.util.extent().fields([d=>d.xlo,d=>d.xhi])(data))
   .yDomain(fc.util.extent().fields([d=>d.low-0.25, d=>d.high+0.25])(data))
   .decorate(function(s) { ... }
```
This works fine. (Actually I like this better for the Y extents, would have used it if I'd known about it.) 

I admit I'm having difficulty getting clear on what the actual proposed change is, but based on the above, it now sounds like the plan is to remove support for strings that represent field names, leaving only the (existing) support for functionals (@chrisprice - is that correct?). I'm less worried about that change, though I'll miss this feature. It's a nice convenience that makes sense, since most of the time that's exactly what you want.

Since this is changing, however, may I make a request that I think fits nicely with this change? How about supporting the full argument list for array.map. It's currently hardcoded in extents to pass only the array items (first of three args) to the passed functional. Could you also pass the index and the array ref to the user's functionals -- as in [d3.min](https://github.com/d3/d3-array/blob/master/src/min.js) ?

**Comment by chrisprice** [comment]
> How about supporting the full argument list for array.map. It's currently hardcoded in extents to pass only the array items (first of three args) to the passed functional. Could you also pass the index and the array ref to the user's functionals -- as in d3.min ?

Frustration with this actually triggered this issue :smile: I've already [added index to the arguments](https://github.com/ScottLogic/d3fc/commit/a46129296d4fe9b1aa04a049ca90b7b7a42ddc57) but it's not yet been released.

**Comment by chrisprice** [comment]
Now released [v10.1.0](https://github.com/ScottLogic/d3fc/releases/tag/v10.1.0)

**Comment by rhewitt** [comment]
> already added index to the arguments

Great! Thanks for that...and for the discussion.

**Comment by chrisprice** [comment]
V1.0.0 is now released on npm, it'll be integrated back into d3fc tomorrow

**Comment by chrisprice** [comment]
So I'm clearly travelling closer to the speed of light than the rest of you :smile: 

https://github.com/ScottLogic/d3fc/releases/tag/v11.0.0
